File: services/auth/src/main/java/com/example/iter/auth/controller/api/UserApiController.java
package com.example.iter.auth.controller.api;

import com.example.iter.auth.controller.api.spec.UserApiSpec;
import com.example.iter.auth.dto.request.AddressUpdateRequest;
import com.example.iter.auth.dto.request.PasswordChangeRequest;
import com.example.iter.auth.dto.request.UserDeleteRequest;
import com.example.iter.auth.dto.request.UserUpdateRequest;
import com.example.iter.auth.dto.response.AddressResponse;
import com.example.iter.auth.dto.response.UserResponse;
import com.example.iter.auth.service.UserAccountService;
import com.example.iter.auth.service.UserAddressService;
import com.example.iter.auth.support.RefreshTokenCookieManager;
import com.example.iter.common.security.CustomUserDetails;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/users")
@RequiredArgsConstructor
public class UserApiController implements UserApiSpec {

    private final UserAddressService userAddressService;
    private final UserAccountService userAccountService;
    private final RefreshTokenCookieManager refreshTokenCookieManager;

    @Override
    @GetMapping("/me")
    public ResponseEntity<UserResponse> getMe(@AuthenticationPrincipal CustomUserDetails principal) {
        return ResponseEntity.ok(userAccountService.getMyProfile(principal.getUser().getId()));
    }

    @Override
    @PatchMapping("/me")
    public ResponseEntity<UserResponse> updateMe(
            @AuthenticationPrincipal CustomUserDetails principal,
            @Valid @RequestBody UserUpdateRequest request
    ) {
        return ResponseEntity.ok(
                userAccountService.updateMyProfile(principal.getUser().getId(), request));
    }

    @Override
    @PatchMapping("/me/password")
    public ResponseEntity<Void> changePassword(
            @AuthenticationPrincipal CustomUserDetails principal,
            @Valid @RequestBody PasswordChangeRequest request
    ) {
        userAccountService.changePassword(principal.getUser().getId(), request);
        return ResponseEntity.noContent().build();
    }

    @Override
    @GetMapping("/me/address")
    public ResponseEntity<AddressResponse> getDefaultAddress(
            @AuthenticationPrincipal CustomUserDetails principal
    ) {
        return ResponseEntity.ok(userAddressService.getDefaultAddress(principal.getUser().getId()));
    }

    @Override
    @PutMapping("/me/address")
    public ResponseEntity<AddressResponse> updateDefaultAddress(
            @AuthenticationPrincipal CustomUserDetails principal,
            @Valid @RequestBody AddressUpdateRequest request
    ) {
        return ResponseEntity.ok(
                userAddressService.updateDefaultAddress(principal.getUser().getId(), request)
        );
    }

    @Override
    @DeleteMapping("/me")
    public ResponseEntity<Void> withdraw(
            @AuthenticationPrincipal CustomUserDetails principal,
            @Valid @RequestBody(required = false) UserDeleteRequest request
    ) {
        userAccountService.withdraw(principal.getUser().getId(), request);
        return ResponseEntity.noContent()
                .header(HttpHeaders.SET_COOKIE, refreshTokenCookieManager.delete().toString())
                .build();
    }
}

File: services/auth/src/main/kotlin/iter/auth/controller/api/UserApiController.kt
package iter.auth.controller.api

import iter.auth.controller.api.spec.UserApiSpec
import iter.auth.dto.request.AddressUpdateRequest
import iter.auth.dto.request.PasswordChangeRequest
import iter.auth.dto.request.UserDeleteRequest
import iter.auth.dto.request.UserUpdateRequest
import iter.auth.dto.response.AddressResponse
import iter.auth.dto.response.UserResponse
import iter.auth.service.UserAccountService
import iter.auth.service.UserAddressService
import iter.auth.support.RefreshTokenCookieManager
import iter.common.security.CustomUserDetails
import jakarta.validation.Valid
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/users")
class UserApiController(
    private val userAddressService: UserAddressService,
    private val userAccountService: UserAccountService,
    private val refreshTokenCookieManager: RefreshTokenCookieManager,
) : UserApiSpec {

    @GetMapping("/me")
    override fun getMe(@AuthenticationPrincipal principal: CustomUserDetails): ResponseEntity<UserResponse> =
        ResponseEntity.ok(userAccountService.getMyProfile(principal.user.id))

    @PatchMapping("/me")
    override fun updateMe(
        @AuthenticationPrincipal principal: CustomUserDetails,
        @Valid @RequestBody request: UserUpdateRequest,
    ): ResponseEntity<UserResponse> =
        ResponseEntity.ok(userAccountService.updateMyProfile(principal.user.id, request))

    @PatchMapping("/me/password")
    override fun changePassword(
        @AuthenticationPrincipal principal: CustomUserDetails,
        @Valid @RequestBody request: PasswordChangeRequest,
    ): ResponseEntity<Void> {
        userAccountService.changePassword(principal.user.id, request)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/me/address")
    override fun getDefaultAddress(
        @AuthenticationPrincipal principal: CustomUserDetails,
    ): ResponseEntity<AddressResponse> =
        ResponseEntity.ok(userAddressService.getDefaultAddress(principal.user.id))

    @PutMapping("/me/address")
    override fun updateDefaultAddress(
        @AuthenticationPrincipal principal: CustomUserDetails,
        @Valid @RequestBody request: AddressUpdateRequest,
    ): ResponseEntity<AddressResponse> =
        ResponseEntity.ok(userAddressService.updateDefaultAddress(principal.user.id, request))

    @DeleteMapping("/me")
    override fun withdraw(
        @AuthenticationPrincipal principal: CustomUserDetails,
        @Valid @RequestBody(required = false) request: UserDeleteRequest?,
    ): ResponseEntity<Void> {
        userAccountService.withdraw(principal.user.id, request)
        return ResponseEntity.noContent()
            .header(HttpHeaders.SET_COOKIE, refreshTokenCookieManager.delete().toString())
            .build()
    }
}
